File: Array & Hash/two_sum.py
# Problem link: https://leetcode.com/problems/two-sum/

# Solution 1:
# Complexity: O(n)

def twoSum(nums, target):
    # a + b  =  target
    # a = target - b
    num_indices = {}

    for i, num in enumerate(nums):
        complement = target - num

        if complement in num_indices:
            return [num_indices[complement], i]

        num_indices[num] = i

    return []

File: Array & Hash/two_sum.cpp
// Problem Link: https://leetcode.com/problems/two-sum/
// Time Complexity O(n).

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {

        unordered_map<int, int> numSet;

        for (int i = 0; i < nums.size(); i++)
        {

            int complement = target - nums[i];

            // if pair found in the map
            if (numSet.find(complement) != numSet.end())
            {
                return {numSet[complement], i};
            }
            // Storing elements as key and its index as value
            numSet[nums[i]] = i;
        }

        return {};
    }
};